addons/adf/src/main/java/org/apache/falcon/adfservice/ADFProviderService.java

```diff
@@ -0,0 +1,376 @@
+        if (requestQueueName == null) {
+            throw new FalconException(AZURE_SERVICEBUS_CONF_PREFIX + AZURE_SERVICEBUS_CONF_REQUEST_QUEUE_NAME
+                    + " property not set in startup properties. Please add it.");
+        }
+        statusQueueName = StartupProperties.get().getProperty(AZURE_SERVICEBUS_CONF_PREFIX
+                + AZURE_SERVICEBUS_CONF_STATUS_QUEUE_NAME);
+        if (statusQueueName == null) {
+            throw new FalconException(AZURE_SERVICEBUS_CONF_PREFIX + AZURE_SERVICEBUS_CONF_STATUS_QUEUE_NAME
+                    + " property not set in startup properties. Please add it.");
+        }
+
+        // init opts
+        opts.setReceiveMode(ReceiveMode.PEEK_LOCK);
+        opts.setTimeout(AZURE_SERVICEBUS_RECEIVEMESSGAEOPT_TIMEOUT);
+
+        // restart handling
+        CurrentUser.authenticate(SUPER_USER);
+        for (EntityType entityType : EntityType.values()) {
+            Collection<String> entities = STORE.getEntities(entityType);
+            for (String entityName : entities) {
+                updateJobStatus(entityName, entityType.toString());
+            }
+        }
+
+        Services.get().<WorkflowJobEndNotificationService>getService(
+                WorkflowJobEndNotificationService.SERVICE_NAME).registerListener(this);
+        adfScheduledExecutorService = new ADFScheduledExecutor(AZURE_SERVICEBUS_REQUEST_HANDLING_THREADS);
+        adfScheduledExecutorService.scheduleWithFixedDelay(new HandleADFRequests(), 0, getDelay(), TimeUnit.SECONDS);
+        LOG.info("Falcon ADFProvider service initialized");
+    }
+
+    private class HandleADFRequests implements Runnable {
+
+        @Override
+        public void run() {
+            String sessionID = null;
+            try {
+                LOG.info("To read message from adf...");
+                ReceiveQueueMessageResult resultQM =
+                        service.receiveQueueMessage(requestQueueName, opts);
+                BrokeredMessage message = resultQM.getValue();
+                if (message != null && message.getMessageId() != null) {
+                    sessionID = message.getReplyToSessionId();
+                    BufferedReader rd = new BufferedReader(
+                            new InputStreamReader(message.getBody()));
+                    StringBuilder sb = new StringBuilder();
+                    String line;
+                    while ((line = rd.readLine()) != null) {
+                        sb.append(line);
+                    }
+                    rd.close();
+                    String msg = sb.toString();
+                    LOG.info("ADF message: " + msg);
+
+                    service.deleteMessage(message);
+
+                    ADFJob.JobType jobType = ADFJob.getJobType(msg);
+                    switch (jobType) {
+                    case REPLICATION:
+                        ADFReplicationJob job = new ADFReplicationJob(msg, sessionID);
```

Comment:
Instead of creating concrete classes for each supported ADFJob type(REPLICATION, HIVE, PIG) user Factory pattern.